src/RpcServer/Fairy.Debugger.DebugInfo.cs

```diff
@@ -0,0 +1,169 @@
+{
+    public partial class RpcServer
+    {
+        public struct SourceFilenameAndLineNum { public string SourceFilename; public uint LineNum;}
+        readonly ConcurrentDictionary<UInt160, HashSet<SourceFilenameAndLineNum>> contractScriptHashToSourceLineNums = new();
+        readonly ConcurrentDictionary<UInt160, Dictionary<uint, SourceFilenameAndLineNum>> contractScriptHashToInstructionPointerToSourceLineNum = new();
+        readonly ConcurrentDictionary<UInt160, HashSet<string>> contractScriptHashToSourceLineFilenames = new();
+        readonly ConcurrentDictionary<UInt160, Dictionary<uint, OpCode>> contractScriptHashToInstructionPointerToOpCode = new();
+        readonly ConcurrentDictionary<UInt160, JObject> contractScriptHashToNefDbgNfo = new();
+        struct DumpNefPatterns
+        {
+            public Regex opCodeRegex = new Regex(@"^(\d+)\s(.*?)\s?(#\s.*)?$");  // 8039 SYSCALL 62-7D-5B-52 # System.Contract.Call SysCall
+            public Regex sourceCodeRegex = new Regex(@"^#\sCode\s(.*\.cs)\sline\s(\d+):\s""(.*)""$");  // # Code NFTLoan.cs line 523: "ExecutionEngine.Assert((bool)Contract.Call(token, "transfer", CallFlags.All, tenant, Runtime.ExecutingScriptHash, neededAmount, tokenId, TRANSACTION_DATA), "NFT payback failed");"
+            public Regex methodStartRegex = new Regex(@"^# Method\sStart\s(.*)$");  // # Method Start NFTLoan.NFTLoan.FlashBorrowDivisible
+            public Regex methodEndRegex = new Regex(@"^# Method\sEnd\s(.*)$");  // # Method End NFTLoan.NFTLoan.FlashBorrowDivisible
+        }
+        readonly DumpNefPatterns dumpNefPatterns = new();
+
+        public static string Unzip(byte[] zippedBuffer)
+        {
+            using var zippedStream = new MemoryStream(zippedBuffer);
+            using var archive = new ZipArchive(zippedStream);
+            var entry = archive.Entries.FirstOrDefault();
+            if (entry != null)
+            {
+                using var unzippedEntryStream = entry.Open();
+                using var ms = new MemoryStream();
+                unzippedEntryStream.CopyTo(ms);
+                var unzippedArray = ms.ToArray();
+                return Encoding.Default.GetString(unzippedArray);
+            }
+            throw new ArgumentException("No file found in zip archive");
+        }
+
+        [RpcMethod]
+        protected virtual JObject SetDebugInfo(JArray _params)
+        {
+            string param0 = _params[0].AsString();
+            UInt160 scriptHash = UInt160.Parse(param0);
+            // nccs YourContractProject.csproj --debug
+            // find .nefdbgnfo beside your .nef contract, and
+            // give me the base64encode(content) of .nefdbgnfo file
+            JObject nefDbgNfo = JObject.Parse(Unzip(Convert.FromBase64String(_params[1].AsString())));
+            contractScriptHashToNefDbgNfo[scriptHash] = nefDbgNfo;
+            // https://github.com/devhawk/DumpNef
+            // dumpnef contract.nef > contract.nef.txt
```

Comment:
Why NEF itself isn't sufficient?

Reply:
> Why NEF itself isn't sufficient?

nef+nefdbgnfo is actually sufficient for a map from source code line num to assembly InstructionPointer, but no source code content is included in nefdbgnfo. DumpNef results have raw source code contents included.

Reply:
DumpNef dumps the contents of the NEF file, intermixing source code if it can find the nefdbgnfo file + the source code of the original contract. I don't understand why you would need the output of DumpNef in order to debug. https://github.com/neo-project/neo-debugger certainly doesn't